newrelic/api/opentelemetry.py

```diff
@@ -0,0 +1,476 @@
+# Copyright 2010 New Relic, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import sys
+from contextlib import contextmanager
+
+from opentelemetry import trace as otel_api_trace
+
+from newrelic.api.application import application_instance
+from newrelic.api.background_task import BackgroundTask
+from newrelic.api.datastore_trace import DatastoreTrace
+from newrelic.api.external_trace import ExternalTrace
+from newrelic.api.function_trace import FunctionTrace
+from newrelic.api.message_trace import MessageTrace
+from newrelic.api.message_transaction import MessageTransaction
+from newrelic.api.time_trace import current_trace, notice_error
+from newrelic.api.transaction import Sentinel, current_transaction
+from newrelic.api.web_transaction import WebTransaction
+from newrelic.core.otlp_utils import create_resource
+
+_logger = logging.getLogger(__name__)
+
+# Attributes that help distinguish span types are
+# sometimes added after span creation and sometimes
+# they are added during span creation.  During the case
+# of the former, we have no way of knowing what sort of
+# library instrumentation to apply, so we have a
+# dictionary of module types as a safeguard.
+#
+# Transactions:
+#   "web": Web Application
+#   "wsgi": WSGI Web Application
+#   "asgi": ASGI Web Application
+#   "other": Background Task
+#   "message": Message Queue
+# Traces:
+#   "other": Function Trace
+#   "db": Database/Datastore
+#   "message": Message Queue
+#
+INSTRUMENTING_MODULE_TYPE = {
```

Comment:
Checking my understanding: this is a mapping of modules to span/transaction types for modules in OTEL that don't have attributes added on span creation that would otherwise help us distinguish what type of instrumentation it is.